Compilation: add missing link file options to cache manifest

mlugg opened the pull request:
I unintentionally closed #23891 while pruning branches; this is the exact same diff.

Resolves: #23869

alexrp reviewed (approved):
I'm not Andrew but FWIW this looks fine to me.

mlugg commented:
Rebased, and tested the `-fentry` case locally -- and at that point I had pretty much written a standalone test case anyway, so I added that in. The test is confirmed as failing on master and passing on this PR.

mlugg commented:
`-fentry` is either broken on our MachO linker or requires something special, so I've just hardcoded the standalone test target to `x86_64-linux`, since I know it works there. That's fine, since we're mainly trying to test the *frontend* there, and in particular the caching mechanism.
